blockmanager.go

```diff
@@ -1026,71 +1026,68 @@ func (b *blockManager) getCheckpointedCFHeaders(checkpoints []*chainhash.Hash,
-				// headers.
-				r.PrevFilterHeader = *curHeader
-				offset := curHeight + 1 - startHeight
-				r.FilterHashes = r.FilterHashes[offset:]
-
-				log.Debugf("Using offset %d for initial "+
-					"filter header range (new prev_hash=%v)",
-					offset, r.PrevFilterHeader)
-			}
-
-			curHeader, err = b.writeCFHeadersMsg(r, store)
-			if err != nil {
-				panic(fmt.Sprintf("couldn't write cfheaders "+
-					"msg: %v", err))
-			}
+			// Add the verified response to our cache.
+			queryResponses[checkPointIndex] = r
 
 			// Then, we cycle through any cached messages, adding
 			// them to the batch and deleting them from the cache.
 			for {
-				// Determine the next checkpoint index we should
-				// process.
-				checkPointIndex += maxCFCheckptsPerQuery
-				if checkPointIndex == uint32(len(checkpoints)) {
-					checkPointIndex = uint32(len(checkpoints)) - 1
-				}
-
-				// We'll also update the current height of the
-				// last written set of cfheaders.
-				curHeight = checkPointIndex * wire.CFCheckptInterval
-
 				// If we don't yet have the next response, then
 				// we'll break out so we can wait for the peers
 				// to respond with this message.
-				r, ok := queryResponses[checkPointIndex]
+				r, ok := queryResponses[currentInterval]
 				if !ok {
 					break
 				}
 
 				// We have another response to write, so delete
 				// it from the cache and write it.
-				delete(queryResponses, checkPointIndex)
+				delete(queryResponses, currentInterval)
 
 				log.Debugf("Writing cfheaders at height=%v to "+
 					"next checkpoint", curHeight)
 
+				// If this is the very first range we've requested, we
+				// may already have a portion of the headers written to
+				// disk.
+				//
+				// TODO(roasbeef): can eventually special case handle
+				// this at the top
+				if bytes.Equal(curHeader[:], initialFilterHeader[:]) {
+					// So we'll set the prev header to our best
+					// known header, and seek within the header
+					// range a bit so we don't write any duplicate
+					// headers.
+					r.PrevFilterHeader = *curHeader
+					offset := curHeight + 1 - startHeight
+					r.FilterHashes = r.FilterHashes[offset:]
+
+					log.Debugf("Using offset %d for initial "+
+						"filter header range (new prev_hash=%v)",
+						offset, r.PrevFilterHeader)
+				}
+
 				// As we write the set of headers to disk, we
 				// also obtain the hash of the last filter
 				// header we've written to disk so we can
 				// properly set the PrevFilterHeader field of
 				// the next message.
-				curHeader, err = b.writeCFHeadersMsg(r, store)
+				curHeader, curHeight, err = b.writeCFHeadersMsg(r, store)
 				if err != nil {
 					panic(fmt.Sprintf("couldn't write "+
 						"cfheaders msg: %v", err))
 				}
+
+				// Update the next interval to write to reflect our
+				// current height.
+				currentInterval = curHeight / wire.CFCheckptInterval
+			}
+
+			// If the current interval is beyond our checkpoints,
+			// we are done.
```

Comment:
Now sure how much of a simplification this commit actually is overall...no context in the commit message.

Reply:
Expanded  commit message.